druid-shell/src/dnd.rs

```diff
@@ -0,0 +1,96 @@
+use std::path::PathBuf;
+
+use crate::backend::dnd as backend;
+use crate::{Counter, FormatId, Modifiers};
+
+use kurbo::Point;
+use piet_common::ImageBuf;
+
+#[derive(Debug)]
+pub enum DragDropAction {
+    Copy,
+    Move,
+}
+
+#[derive(Debug)]
+pub struct DragData(backend::DragData);
+
+#[derive(Debug, Clone)]
+pub struct DropEvent {
+    pub modifiers: Modifiers,
+    pub position: Point,
+}
+
+#[derive(Clone)]
+pub struct DropContext(pub(crate) backend::DragDropContext);
+
+/// A unique identifier for a drag drop session.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct DragDropId(u64);
+
+impl DragDropId {
+    /// Create a new, unique DragDropId
+    pub(crate) fn next() -> DragDropId {
+        static COUNTER: Counter = Counter::new();
+        DragDropId(COUNTER.next())
+    }
+}
+
+impl DragData {
+    pub fn new() -> Self {
+        DragData(backend::DragData::new())
+    }
+
+    /// Add a data format
+    pub fn add(&mut self, format: FormatId, data: Vec<u8>) {
+        self.0.add(format, data)
+    }
+
+    /// default: true
+    pub fn copyable(&mut self, allowed: bool) {
+        self.0.copyable(allowed)
+    }
+
+    /// default: false
+    pub fn movable(&mut self, allowed: bool) {
+        self.0.movable(allowed)
+    }
+
+    pub fn cursor_image(&mut self, image: ImageBuf) {
+        self.0.cursor_image(image)
+    }
+
+    pub fn files(&mut self, files: Vec<PathBuf>) {
+        self.0.files(files)
+    }
+}
+
+impl DropContext {
```

Comment:
user receives this. When `set_action` is called, it sets the current action. 
for example in case of `drag_motion`, `set_action(DragDropAction::Copy)` if Ctrl is held

Reply:
done by the source: `DragData::cursor_image` (needs a better name)